src/server/ua_services_session.c

```diff
@@ -415,16 +415,27 @@ Service_CreateSession(UA_Server *server, UA_SecureChannel *channel,
 
 static UA_StatusCode
 checkSignature(const UA_Server *server, const UA_SecurityPolicy *securityPolicy,
-               void *channelContext, const UA_ByteString *serverNonce,
-               const UA_SignatureData *signature) {
+               void *channelContext, const UA_ByteString *serverNonce,const UA_UserTokenPolicy *userTokenPolicy,
+               const UA_ExtensionObject *userToken, const UA_SignatureData *signature) {
     /* Check for zero signature length */
     if(signature->signature.length == 0)
         return UA_STATUSCODE_BADAPPLICATIONSIGNATUREINVALID;
 
     if(!securityPolicy)
         return UA_STATUSCODE_BADINTERNALERROR;
 
+    /* Server certificate */
     const UA_ByteString *localCertificate = &securityPolicy->localCertificate;
+
+    if(userTokenPolicy != NULL && userToken != NULL) {
+        if(userTokenPolicy->tokenType == UA_USERTOKENTYPE_CERTIFICATE) {
+            UA_X509IdentityToken *userIdentityToken = (UA_X509IdentityToken *)
+                userToken->content.decoded.data;
+            /* Client certificate */
+            localCertificate = &userIdentityToken->certificateData;
+        }
+    }
+
```

Comment:
Remove. The data that is signed is always the server certificate plus the last server nonce.
See: https://reference.opcfoundation.org/Core/Part4/v104/docs/5.6.3
The data to sign is created by appending the last [serverNonce](https://reference.opcfoundation.org/search/17?t=serverNonce)to the [serverCertificate](https://reference.opcfoundation.org/search/17?t=serverCertificate)specified in the [CreateSession](https://reference.opcfoundation.org/search/17?t=CreateSession)response.